lib/ws-channel.js

```diff
@@ -0,0 +1,250 @@
+  assert(this._token);
+  this._uri = null; // TODO(bert) set this
+  this._attach(websocket);
+
+  return this;
+};
+
+
+WebsocketChannel.prototype._attach = function _attach(websocket) {
+  this._websocket = websocket;
+
+  websocket.on('message', this._handleMessage.bind(this));
+  websocket.on('error', this._handleError.bind(this));
+
+  if (websocket.readyState === Websocket.CONNECTING)
+    websocket.once('open', this._onOpen.bind(this));
+  else
+    this._onOpen();
+
+  this._flushSendQueue();
+};
+
+
+WebsocketChannel.prototype.close = function close(callback) {
+  if (this._websocket)
+    this._websocket.close();
+
+  if (this._sendQueue.length === 0) {
+    process.nextTick(callback);
+
+  } else {
+    var error = new Error('Messages were discarded');
+    process.nextTick(function() {
+      callback(error);
+    });
+  }
+
+  return this;
+};
+
+
+WebsocketChannel.prototype.notify = function notify(message) {
+  var packet = {
+    type: NOT,
+    seqno: this._seqno++,
+    data: message
+  };
+
+  this._send(packet);
+};
+
+
+WebsocketChannel.prototype.request = function request(message, callback) {
+  var seqno = this._seqno++;
+  this._callbacks[seqno] = callback;
+  this._callbacksPending++;
+
+  // Reference the socket (even if was referenced) if we're waiting for a
+  // response.
+  if (this._callbacksPending == 1 && this._socket && this._unref)
+    this._socket.ref();
+
+  var packet = {
+    type: REQ,
+    seqno: seqno,
+    data: message
+  };
+
+  this._send(packet);
+};
+
+
+WebsocketChannel.prototype.address = function address() {
+  return this.uri && url.parse(this.uri);
+};
+
+
+WebsocketChannel.prototype.getToken = function getToken() {
+  return this._token;
+};
+
+
+WebsocketChannel.prototype.unref = function unref() {
+  this._unref = true;
+  if (this._socket && this._callbacksPending === 0)
+    this._socket.unref();
+};
+
+
+WebsocketChannel.prototype._onOpen = function() {
+  this._socket = this._websocket._socket;
+
+  if (this._unref && this._callbacksPending === 0)
+    this._socket.unref();
+
+  this._flushSendQueue();
+};
+
+
+WebsocketChannel.prototype._send = function(message) {
+  this._sendQueue.push(JSON.stringify(message));
+  if (this._isOpen())
+    this._flushSendQueue();
+};
+
+
+WebsocketChannel.prototype._handleMessage = function(s) {
+  var self = this;
+
+  try {
+    var packet = JSON.parse(s);
+  } catch (err) {
+    return this._handleError(err);
+  }
+
+  switch (packet.type) {
+    case NOT:
+      return this._onNotification(packet.data);
+
+    case REQ:
+      return this._onRequest(packet.data, sendReply);
+
+    case RSP:
+      var callback = this._callbacks[packet.seqno];
+      if (!callback)
+        return this._handleError(new Error('Unrecognized response sequence number'));
+
+      delete this._callbacks[packet.seqno];
+      this._callbacksPending--;
+
+      // Unref the socket if the channel was unref'ed and we're no longer
+      // awaiting any responses.
```

Comment:
I note you are going through a lot of trouble to do this (here, `WebsocketChannel.request()`, etc.), but I don't understand why. If pm has no refed handles or requests, and it has unrefed the WebsocketChannel, then it should exit. It has indicated it doesn't want to be kept alive, even if there are pending responses.

This appears to be a "graceful close" mechanism, almost, wherein the unref() causes the Websocket to be unrefed only when there is nothing outstanding. I don't think supervisor or pm needs or would use such a mechanism, and I think I went through some effort to ensure and test, see test-server-unref.js, that the control channels don't keep either end alive.

Did you find something that needed this feature?